cogs5e/models/automation/effects/damage.py

```diff
@@ -91,9 +96,20 @@ def run(self, autoctx):
         # Disable critical damage in saves (#1556)
         in_crit = (autoctx.in_crit or crit_arg) and not (nocrit or autoctx.in_save)
         if in_crit:
-            dice_ast = d20.utils.tree_map(utils.crit_mapper, dice_ast)
+            if crit_damage_type == CritDamageType.MAX_ADD:
+                dice_ast = d20.utils.tree_map(utils.max_add_crit_mapper, dice_ast)
+            elif crit_damage_type == CritDamageType.DOUBLE_ALL:
+                dice_ast.roll = d20.ast.BinOp(d20.ast.Parenthetical(dice_ast.roll), "*", d20.ast.Literal(2))
+            elif crit_damage_type == CritDamageType.DOUBLE_DICE:
+                dice_ast = d20.utils.tree_map(utils.double_dice_crit_mapper, dice_ast)
+            else:
+                dice_ast = d20.utils.tree_map(utils.crit_mapper, dice_ast)
             if critdice and not autoctx.is_spell:
-                utils.critdice_tree_update(dice_ast, int(critdice))
+                if crit_damage_type in (CritDamageType.DOUBLE_ALL, CritDamageType.DOUBLE_DICE):
+                    crit_ast = utils.crit_dice_gen(dice_ast, critdice)
+                    dice_ast.roll = d20.ast.BinOp(dice_ast.roll, "+", crit_ast)
+                else:
```

Comment:
Actually, is this code necessary? `utils.critdice_tree_update` should still work right for these crit types, right? Given an example damage of `1d6 + 3`, they get modified to:

- max_add: `1d6 + 3 + 6`
- double_all: `(1d6 + 3) * 2`
- double_dice: `(1d6) * 2 + 3` (BTW, you can probably remove the parenthetical from this, it seems a bit redundant)
- normal: `2d6 + 3`

and the tree updater should correctly go along the left branch and add *critdice* rolls to the leftmost roll in each of these cases.

Reply:
In the case of Double All and Double Total, the general consensus seems to be that you don't double EXTRA crit dice, only the originating

Reply:
So for the example I gave above, assuming `critdice=1`, it would turn into:
- max_add: `2d6 + 3 + 6`
- double_all: `(1d6 + 3) * 2 + 1d6`
- double_dice: `(1d6) * 2 + 3 + 1d6`
- normal: `3d6 + 3`
? Gross.

Reply:
Yuuup. I originally had it just add like normal but was called out in #contrib and ddb mod chat for that being wrong haha